src/cargo/core/resolver/encode.rs

```diff
@@ -68,36 +163,109 @@ impl EncodableResolve {
                     Some(&source) => PackageId::new(&pkg.name, &pkg.version, source)?,
                 };
 
+                // If a package has a checksum listed directly on it then record
+                // that here, and we also bump our version up to 2 since V1
+                // didn't ever encode this field.
+                if let Some(cksum) = &pkg.checksum {
+                    version = ResolveVersion::V2;
+                    checksums.insert(id, Some(cksum.clone()));
+                }
+
                 assert!(live_pkgs.insert(enc_id, (id, pkg)).is_none())
             }
             live_pkgs
         };
 
-        let lookup_id = |enc_id: &EncodablePackageId| -> Option<PackageId> {
-            live_pkgs.get(enc_id).map(|&(id, _)| id)
-        };
+        // When decoding a V2 version the edges in `dependencies` aren't
+        // guaranteed to have either version or source information. This `map`
+        // is used to find package ids even if dependencies have missing
+        // information. This map is from name to version to source to actual
+        // package ID. (various levels to drill down step by step)
+        let mut map = HashMap::new();
+        for (id, _) in live_pkgs.values() {
+            map.entry(id.name().as_str())
+                .or_insert(HashMap::new())
```

Comment:
Premature optimization: `HashMap::new()` doesn't heap-allocate IIRC but maybe it'd be even better to use the lazy `or_default()` instead?

Reply:
Heh yeah I wouldn't worry too much about this, Cargo has a huge number of points it allocates unnecessarily, but this won't ever show up in any profile even if `HashMap::new()` did indeed allocate

Reply:
Makes sense, just spotted that as I browsed the diff =)